---
title: "Petr Mockup"
author: "Brian S. Yandell"
date: "3/17/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The motivation is to reduce memory usage. Matt told us that memory is his greatest obstacle, when running QTL browsers because he cannot add more `R/plumber` workers. Also, if you have a cheap Windows laptop you cannot even open many DO `.RData` projects simply because haplotype probability object is too big.
 
The solution would be to store data in some database and load just a slice you need at a time. I experimented with `SQLite` but it was too slow. Then I try feather and it run fine (R script attached â€“ crunching 1000 SNPs at a time). It just does not use `qtl2scan` effectively because the package is not prepared for that.

Task: refactor this Rmd to use DOex and run some examples.

```{r}
library(dplyr)
library(qtl2geno)
library(qtl2scan)
library(feather)
```

```{r}
if(!file.exists("DOex.zip")) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
} else {
  dirpath <- "."
}
file <- file.path(dirpath, "DOex.zip")
DOex <- read_cross2(file)
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Write feather database of pr.

```{r}
fpr <- feather_genoprob(pr, "pr.feather", "feather")
fapr <- feather_genoprob(apr, "apr.feather", "feather")
```

```{r}
dim(fapr[["X"]])
```

The following does not work well. Yet.

```{r}
try(dim(fapr$X))
```

```{r}
dim(subset(fapr, ind=1:20, chr=c("2","3"))[["3"]])
```

Start of work on `cbind.feather_genoprob`. Major step is to create new feather when combining objects from different feathers. But don't want to propagate, and want to think about where to put the feather file, say in same directory as first object. But then need to work out naming conventions.

```{r}
knitr::knit_exit()
```

Set up batches for reading

```{r}
Nbatches = 45
Nmar <- dim(apr[[1]])[3]
# make SNPs split into Nbatches
batches <-  split(1:Nmar, ceiling(seq_along(1:Nmar)/Nbatches))
```

Read pr many times.

```{r}
for(i in seq)
```

LOD object to be produced

```{r}
result <- NULL

for (i in 1:length(batches)) {

  # fetch data from the database
  tmp <- read_feather(db.file, columns = batches[[i]])
  print(i)
  
  # dimnames harcoded, ADD ASSERTIONS LATER 
  tmp.snps <- names(tmp)
  tmp.probs <- as.array(as.matrix(tmp))
  dim(tmp.probs) <- c(nrow(tmp.probs) %/% 8, 8, ncol(tmp.probs))
  dimnames(tmp.probs) <- list(annot.samples$Mouse.ID, LETTERS[1:8], tmp.snps)
  
  # convert to qtl2 format
  probs <- probs_doqtl_to_qtl2(tmp.probs, snps[tmp.snps,], pos_column = "bp")
  chrs <- unique(snps[tmp.snps,"chr"])
  
  tmp.fit <- scan1(genoprobs=probs, 
               kinship=Glist[chrs],
               pheno=expr.mrna[,idx], 
               addcovar=addcovar, 
               cores=1, 
               reml=TRUE)
  
  # save LOD scores to results
  result <- rbind(result, tmp.fit$lod)
}
```

