# subset feather_genoprob objects

#' Subsetting genotype probabilities
#'
#' Pull out a specified set of individuals and/or chromosomes from
#' the results of \code{\link{feather_genoprob}}.
#'
#' @param x Genotype probabilities as output from \code{\link{feather_genoprob}}.
#' @param ind A vector of individuals: numeric indices, logical
#' values, or character string IDs
#' @param chr A vector of chromosomes: logical values, or character
#' string IDs. Numbers are interpreted as character string IDs.
#' @param ... Ignored.
#'
#' @return The input genotype probabilities, with the selected
#' individuals and/or chromsomes.
#'
#' @export
#' @export subset.feather_genoprob
#' @method subset feather_genoprob
#' 
#' @keywords utilities
#'
#' @examples
#' library(qtl2geno)
#' grav2 <- read_cross2(system.file("extdata", "grav2.zip", package="qtl2geno"))
#' \dontshow{grav2 <- grav2[1:8,c(1,2)]}
#' pr <- calc_genoprob(grav2)
#' fpr <- feather_genoprob(pr, "my.feather")
#' # keep just individuals 1:5, chromosome 2
#' prsub <- fpr[1:5,2]
#' # keep just chromosome 2
#' prsub2 <- fpr[,2]
subset.feather_genoprob <-
    function(x, ind=NULL, chr=NULL, ...)
{
    if(is.null(ind) && is.null(chr))
        stop("You must specify either ind or chr.")
    
    chrID <- names(x)
    n_chr <- length(chrID)
    if(!is.null(chr)) {
        if(is.logical(chr)) {
            if(length(chr) != n_chr) {
                stop("length(chr) [", length(chr), "] != no. chr in x [",
                     n_chr, "]")
            }
            chr <- chrID[chr] # convert to character strings
        }
        else {
            chrindex <- match(chr, chrID)
            if(any(is.na(chrindex))) {
                stop("Some chr not found: ",
                     paste(chr[is.na(chrindex)], collapse=", "))
            }
            chr <- chrID[chrindex] # character strings
        }
        if(length(chr) == 0)
            stop("Must retain at least one chromosome.")
      
        cl <- class(x)
        class(x) <- "list"
      
        attr_to_sub <- c("is_x_chr")
        attr_to_keep <- c("crosstype", "alleles", "alleleprobs")
        x_attr <- attributes(x)
        x_attrnam <- names(x_attr)
        x <- x[chr]
      for(a in attr_to_sub) {
        if(a %in% x_attrnam) {
          attr(x, a) <- x_attr[[a]][chr]
        }
      }
      for(a in attr_to_keep) {
        if(a %in% attr_to_keep)
          attr(x, a) <- x_attr[[a]]
      }
      class(x) <- cl
    }
    
    indID <- rownames(x[[1]])
    
    n_ind <- length(indID)
    if(!is.null(ind)) {
        if(is.logical(ind)) {
            if(length(ind) != n_ind) {
                stop("length(ind) [", length(ind), "] != no. ind in x [",
                     n_ind, "]")
            }
            ind <- indID[ind] # convert to character strings
        }
        else if(is.numeric(ind)) {
            if(any(ind < 1 || ind > n_ind))
                stop("Numeric ind out of allowed range [1 - ", n_ind, "]")
            ind <- indID[ind] # convert to character strings
        }
        else {
            indindex <- match(ind, indID)
            if(any(is.na(indindex))) {
                stop("Some individuals not found: ",
                     paste(ind[is.na(indindex)], collapse=", "))
            }
            ind <- indID[indindex]
        }
        if(length(ind) == 0)
            stop("Must retain at least one individual.")
      
        cl <- class(x)
        class(x) <- "list"
      
        for(i in names(x)) # loop over chromosomes
          x[[i]] <- x[[i]][ind,,,drop=FALSE]
      
        class(x) <- cl
      
    }

    x
}

#' @export
#' @export [.feather_genoprob
#' @method [ feather_genoprob
#' 
#' @rdname subset.feather_genoprob
`[.feather_genoprob` <-
    function(x, ind=NULL, chr=NULL)
    subset(x, ind, chr)
